File: jee/src/test/scala/e2e/E2ETestBase.scala
package e2e

import org.junit.{Test,Before,After,BeforeClass,AfterClass}

class E2ETestBase extends org.scalatest.selenium.Chrome with com.typesafe.scalalogging.slf4j.LazyLogging {
  // http://www.scalatest.org/user_guide/using_selenium
  def root:String = "http://localhost:%d/".format(E2ETestBase.port)

  @Before def before():Unit = go to root
  @After def after():Unit = close
}

object E2ETestBase {
  var server:org.eclipse.jetty.server.Server = _
  var port:Int = _

  modules.config.loadConfig("standalone")

  val driverPath =
    if (System.getProperty("os.name").startsWith("Windows")) "bin/chromedriver.exe"
    else "bin/chromedriver"

  if (new java.io.File(driverPath).isFile) {
    System.setProperty("webdriver.chrome.driver", driverPath)
  }

  @BeforeClass
  def beforeClass():Unit = {
    val (server, port) = modules.webapp.startServer
    this.server = server
    this.port = port
  }

  @AfterClass def afterClass():Unit = server.stop
}

class SampleE2ETest extends E2ETestBase {
  @Test def testBar():Unit = {
    go to root + "markdown/"
  }
}

File: jee/src/test/java/e2e/E2ETestBase.java
package e2e;

import org.junit.Test;
import org.junit.Before;
import org.junit.After;
import org.junit.BeforeClass;
import org.junit.AfterClass;
import org.eclipse.jetty.server.Server;
import org.openqa.selenium.chrome.ChromeDriver;

public class E2ETestBase {
  static {
    modules.config.JavaCompat.loadConfig("standalone");
    String driverPath;
    if (System.getProperty("os.name").startsWith("Windows")) {
      driverPath = "bin/chromedriver.exe";
    } else {
      driverPath = "bin/chromedriver";
    }
    if (new java.io.File(driverPath).isFile()) {
      System.setProperty("webdriver.chrome.driver", driverPath);
    }
    driver = new ChromeDriver();
  }

  static Server server;
  static int port;
  protected static ChromeDriver driver;

  @BeforeClass public static void beforeClass()
  {
    scala.Tuple2<org.eclipse.jetty.server.Server,Integer> tpl = modules.jetty.JavaCompat.startServer();
    server = tpl._1;
    port = tpl._2;
  }

  @AfterClass public static void afterClass() throws Exception
  {
    server.stop();
  }

  protected String root() { return "http://localhost:" + port + "/"; }

  @Before public void before() { driver.get(root()); }
  @After public void after() { driver.close(); }
}
